### PR TITLE
[WIP] Implement wasi-http (requests) for wazero

### PR DESCRIPTION
**Pull request by brendandburns**
This is a working pull request, but it's not ready for submission, it is a work in progress to determine where this should land.

It implements the [wasi-http](https://github.com/WebAssembly/wasi-http) specification for HTTP requests from web assembly, but it's pretty hacky just to show what is possible.

I'm sending it now to determine if there is interest in merging this into the repo.

If there is interest, then we can discuss the final structure and I'll clean up the code (a lot)

If there's not interest, I'll find another home for it.

Feedback very welcome, thanks!

### CODE REVIEW

**Comment by brendandburns** [comment]
I'm attaching a sample http wasm file from https://github.com/dev-wasm/dev-wasm-c/tree/main/http

If you want to try it out, build in this PR, then run

`wazero run main.wasm`
[main.wasm.zip](https://github.com/tetratelabs/wazero/files/11739611/main.wasm.zip)

**Review by achille-roussel — COMMENTED** [review]
Hello @brendanburns, thanks a lot for the contribution, we love seeing those kinds of initiatives 🙌 

I left a couple of questions on the diff, but besides that I think you're touching on the right point when asking where would be a good place to host this code.

At this time, the wasi-http and streams (which I believe intend to be part of WASI preview 2?) seem to be too experimental to be merged into the main wazero repository. We have been focused on having a robust syscall-like layer based on WASI preview 1, and currently finalizing the addition of non-blocking capabilities to enable a large class of concurrent application to run in Wazero. Because Wazero has reached v1.0, we also commit to the long-term maintenance of the APIs we introduce.

Something you might be interested in as a potential home for this code is https://github.com/stealthrocket/wasi-go. We started that project to have a place to publish those types of experimental wazero host modules, with less concerns about having to introduce breaking changes or portability across languages or operating systems.

Alternatively, this can be hosted in its own repository, and we would be happy to add a link in the Wazero community catalog we maintain at https://github.com/tetratelabs/wazero/blob/main/site/content/community/users.md so that we help users find the code that they are interested in.

Let me know which approach you think would make most sense!

File: imports/wasi_http/http.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+func Malloc(ctx context.Context, m api.Module, size uint32) (uint32, error) {
+	malloc := m.ExportedFunction("cabi_realloc")
+	result, err := malloc.Call(ctx, 0, 0, 4, uint64(size))
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+	return uint32(result[0]), err
+}
```
[the remark this answers is not in the document]
Comment (brendandburns) [reply imports/wasi_http/http.go:240]: Regarding `cabi_realloc` it's the way that `wasmtime` works, there's some discussion here:

https://github.com/bytecodealliance/preview2-prototyping/issues/99

and also some discussion here:

https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md

search for `realloc`

File: imports/wasi_streams/read.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+var streamsRead = newHostMethod("read", streamReadFn, []api.ValueType{i32, i64, i32}, "a", "b", "c")
+
+func streamReadFn(ctx context.Context, mod api.Module, params []uint64) {
+	data, err := wasi_http.ResponseBody()
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+	ptr_len := uint32(len(data))
+	data = append(data, 0)
+	ptr, err := wasi_http.Malloc(ctx, mod, ptr_len)
```
[the remark this answers is not in the document]
Comment (brendandburns) [reply imports/wasi_streams/read.go:20]: There's some discussion of this here:

https://github.com/WebAssembly/wasi-http/issues/7

File: imports/wasi_streams/read.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import (
+	"context"
+	"log"
+
+	"github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/imports/wasi_http"
+)
+
+var streamsRead = newHostMethod("read", streamReadFn, []api.ValueType{i32, i64, i32}, "a", "b", "c")
```
[the remark this answers is not in the document]
Comment (brendandburns) [reply imports/wasi_streams/read.go:11]: yeah, wasi-http depends on `poll` in wasi:

File: imports/default_http/http.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/internal/wasip1"
+	"github.com/tetratelabs/wazero/internal/wasm"
+)
+
+// ModuleName is the module name WASI functions are exported into.
+//
+// See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md
```
[the remark this answers is not in the document]
Comment (brendandburns) [reply imports/default_http/http.go:15]: oops, sorry that is a cut/paste left over. I stole a bunch of this code from the wasi implementation :)

**Comment by brendandburns** [comment]
@achille-roussel thanks for the quick response. Regarding the wasi-http API it is documented here:

https://github.com/WebAssembly/wasi-http

and has been implemented in the `wasmtime` runtime from the bytecode alliance (`9.0.x`)

It's your call whether that is sufficiently stable for `wazero` though I would argue that it is in the ballpark of stability for wasi.

But I'm happy to pull it out as a separate module. I mainly want this for integration with the [dapr](https://dapr.io) project which relies on wazero.

Let me know what you think.

**Review by codefromthecrypt — COMMENTED** [review]
Thanks for getting this started, it is interesting to have an http client impl and seems will continue as things like the server stabilizes.

I concur with @achille-roussel that trying to develop here wouldn't work out well for reasons mentioned. It is also the case generally PRs longer than a week are a bug (we try to keep the queue clean and things close quickly). Something like this would need a lot of work and could distract folks from the daily issues such as waspi1 stuff which is nearly every day still.

Since Achille offered [go-wasi](https://github.com/stealthrocket/wasi-go), I would take it if I were you ;) You get access to very strong developers who also work on Go. Independent is another route, but basically wasi-go is an alternate CLI so easier to test.

This would help this mature with norms of the wazero ecosystem including tests, benchmarks etc. I can't comment dapr adoption but I think the first thing to do is become stable both in spec and impl and working with wasmtime and also wazero seems a good way towards that, IMHO.

**Comment by brendandburns** [comment]
@achille-roussel thanks for the offer to host this in wasi-go, I will refactor this PR to target that repo instead and see what that looks like.

I'll leave this PR open for another day or two in case there is additional discussion, but I'll close it once I open a PR on wasi-go.

**Comment by achille-roussel** [comment]
Sounds good!

If you have any questions about wasi-go that are better suited for a short conversation feel free to reach out on the Gopher Slack.

**Comment by brendandburns** [comment]
As discussed, closing in favor of https://github.com/stealthrocket/wasi-go/pull/56

**Pull request closed without merging**